### ISSUE TITLE
Possible memory corruption?

### ISSUE THREAD

**Issue by Lemonymous** [godotengine/godot#70235]
### Godot version

4.0.dev (30800d20f452015c054dc7d1d7cfd38362fe28ff)

### System information

Windows 10, Forward+, NVIDIA GeForce GTX 1050 Ti (31.0.15.1694)

### Issue description

I actually don't understand what is happening with this issue. I have only narrowed down an error to a small bit of code.

The code below will throw this error:
```
Invalid type in GDScript utility function '<unknown function>'. Cannot convert argument 2 from Nil to float.
```

I can get rid of the error by doing either of the following:
- if I don't do the calculation for the upper range in the loop. lim+1 -> lim
- if I don't set the variable `a` (which is `null` since `arr` doesn't have any values initialized)
- if I remove the for loop.
- if I change Array[int] to Array
- if I check the `typeof(lim)` before the for loop

Every separate part of the code seems to be working on their own. But combined, it throws the error.

### Steps to reproduce

The following code will produce the error.
```gdscript
func _ready():
	var lim:int = 1
	var arr:Array[int] = []
	arr.resize(1)
	var a = arr[0]
	
	for x in range(1, lim+1):
		pass
```

### Minimal reproduction project

A project with only the code snippet above.
[Possiible memory corruption.zip](https://github.com/godotengine/godot/files/10252961/Possiible.memory.corruption.zip)

**Comment by adamscott** [comment]
Confirmed on master (0bb94df).
Weird. The `lim + 1` argument is considered to be `Variant::Type::NIL`, which trips `range()`

**Comment by adamscott** [comment]
The `range()` method has nothing to do with the issue.

```gdscript
func _ready():
	var lim:int = 1
	var arr:Array[int] = []
	arr.resize(1)
	var a = arr[0]
	print(lim + 1)  # prints "<null>"
```

**Comment by adamscott** [comment]
This seems to be linked with the `Array[int]` typing. Using `PackedInt{32,64}Array` doesn't make `lim + 1 == null`.
I even tried `Array[float]` and `lim + 1 == 2`, as intended.

**Comment by adamscott** [comment]
@vonagam Does your PR #69248 fix this issue? You seem to know a lot about array typing and their quirks.

**Comment by vonagam** [comment]
Hm. No that PR does not fix this issue. There are two problems here.

First one (which is a separate but clear one) is in the issue description:
> (which is null since arr doesn't have any values initialized)

In resize of packed arrays (means `Vector<T>`) they do `resize_zeroed` and there is template logic that if type `T` needs initialization than it is applied.

In Array it is a simple call to `resize` so not even the type of new variants is set. So it is always nulls, which is bad since `Array[int]` should not allow nulls.

But as you said that if you change the type of array in question from `Array[int]` to `Array[float]` or, for example, to `Array[bool]` then the issue is gone. So there is a second problem (the main one), which I have no idea about.

**Comment by chrisb123** [comment]
var class_float:float = 1.111
var class_int:int = 1
var arr_float:Array[float]
var arr_int:Array[int]

func _ready():
	arr_float.resize(1)
	prints("read array",arr_float[0])
	
	var method_float:float = 2.222
	var method_int:int = 2
	
	prints("fail",method_float+1)
	prints("works",method_int+1)
	prints("fail",class_float+1)
	prints("works",class_int+1)
	stuff()
	
func stuff():
	arr_int.resize(1)
	prints("read array",arr_int[0])
	
	var method_float:float = 3.333
	var method_int:int = 3
	
	prints("works",method_float+1)
	prints("fail",method_int+1)
	prints("works",class_float+1)
	prints("fail",class_int+1)

**Comment by adamscott** [comment]
@chrisb123 You can add `gdscript` syntax highlight by adding this to the code block:

\`\`\`gdscript
[CODE]
\`\`\`

**Comment by Lemonymous** [comment]
Possibly related:

The following causes no issue
```
var arr = PackedInt32Array()
arr.resize(1)
```
This hangs the project
```
var arr:PackedInt32Array = PackedInt32Array()
arr.resize(1)
```
Same issue with all the following types
```
PackedByteArray
PackedColorArray
PackedFloat32Array
PackedFloat64Array
PackedInt32Array
PackedInt64Array
PackedStringArray
PackedVector2Array
PackedVector3Array
```

**Comment by adamscott** [comment]
@Lemonymous Please update your build. You're 166 commits late. This has been fixed by #71228.

**Comment by adamscott** [comment]
The issue still exists on master (60d0317), just tested.

**Comment by Lemonymous** [comment]
Thanks adamscott. I retested in `v4.0.beta.custom_build [60d031777]`

First the code in https://github.com/godotengine/godot/issues/70235#issuecomment-1382838032 which I thought was related. You are indeed right that it has been fixed.

I then retested the issue in the opening post with the code provided by chrisb123 https://github.com/godotengine/godot/issues/70235#issuecomment-1356682309

The results were almost the same as in that post, but with one exception. Seems like using `typeof` no longer "fixes" it, so the last line also prints null. So something seems to have changed since that time.
```GDScript
...
typeof(class_int) #this fixes it
prints("works",class_int+1) #this now prints "works <null>" which I don't believe was the case before.
```
